### DISCUSSION TITLE
Help me understand why we use 2 different invariants when adding liquidity & swapping tokens ?

### DISCUSSION THREAD

**Discussion by TilakMaddy** [Cyfrin/security-and-auditing-full-course-s23#49] [Q&A] [answered]
Purely by looking at the source code for `TSwapPool.sol` I infer the following:

For swapping tokens we maintain `x * y = k`

but when it comes to adding liquidity, we try to maintain  `x / y = C`

why do we do that ? and how does the whole math workout?

**Comment by PatrickAlphaC** [comment]
Where do we do that?

And a few times we do some advanced algebra to derive equations from the y * x = k equation.

**Reply by TilakMaddy** [reply, marked as answer]
I got my question answered after going through the formal verification paper [here](https://github.com/runtimeverification/verified-smart-contracts/blob/master/uniswap/x-y-k.pdf) 

`x * y = k` is not true throughout the lifetime of the contract. Otherwise, you could never add extra money (for either token - x or y) without removing some from the other (how else would you have a product of two non-negative numbers be constant, right?). Therefore when it comes to adding liquidity, the value of `x * y` is expected to strictly increase.

The invariant in this situation is the ratio `e:t:l` (e -> ether, t -> exchange token, l -> total liquidity amount). Say you want to provide `Δe` to liquidity pool, in order to preserve the `e:t:l` ratio they also expect you to provide `alpha * t` to `t` and,  in turn you will be minted `alpha * l`. 

where `alpha = e / Δe` 

This is enforced in `TSwap.sol` [here](https://github.com/Cyfrin/5-t-swap-audit/blob/f4337615b7eada3d871378ff8b70af08277d3dec/src/TSwapPool.sol#L429C6-L429C6)

However, when it comes to swapping, the invariant is `x * y = k`

**Reply by shobhit-gupta** [reply]
@TilakMaddy Thank you for the explanation & the link to the verification paper.
